src/dbnode/storage/flush.go

```diff
@@ -270,25 +296,103 @@ func (m *flushManager) indexFlush(
 
 	m.setState(flushManagerIndexFlushInProgress)
 	var (
-		start    = m.nowFn()
-		multiErr = xerrors.NewMultiError()
+		start     = m.nowFn()
+		nsResults = make([][]time.Time, len(namespaces))
+		multiErr  = xerrors.NewMultiError()
 	)
-	for _, ns := range namespaces {
+	for i, ns := range namespaces {
 		var (
 			indexOpts    = ns.Options().IndexOptions()
 			indexEnabled = indexOpts.Enabled()
 		)
 		if !indexEnabled {
 			continue
 		}
-		multiErr = multiErr.Add(ns.FlushIndex(indexFlush))
+		blockStarts, err := ns.FlushIndex(indexFlush)
+		if err != nil {
+			multiErr = multiErr.Add(err)
+		}
+		nsResults[i] = blockStarts
 	}
 	multiErr = multiErr.Add(indexFlush.DoneIndex())
 
+	// Only write out empty index snapshots if the commit log rotation
+	// was successful we don't mind index flush errors since both
+	// index flush and writing empty index snapshots is best effort.
+	if rotateCommitlogSuccess {
+		multiErr = multiErr.Add(m.writeEmptyIndexSnapshots(
+			namespaces,
+			nsResults,
+			startTime,
+			rotatedCommitlogID,
+			snapshotID,
+		))
+	}
+
 	m.metrics.indexFlushDuration.Record(m.nowFn().Sub(start))
 	return multiErr.FinalError()
 }
 
+// NB(bodu): After a successful warm block flush, we must write an empty snapshot to disk.
+// If the DB node crashes before next snapshot is written to disk and we bootstrap the warm snapshot
+// into the now "sealed" block those warm bootstrapped segments will never get evicted from memory.
+// Since we are doing this after a successful flush, there is a small window where we can crash and
+// run into the above case. However, we'd rather deal w/ a one time extra mem than long bootstrap times
+// if we write an empty snapshot to disk and the node crashes before a successful warm flush.
+func (m *flushManager) writeEmptyIndexSnapshots(
```

Comment:
@notbdu as discussed, is it possible at all to avoid writing empty index snapshots now we are tracking everything with snapshot ID?

Reply:
So the issue of potentially bootstrapping a "stale" snapshot if the node crashes between a successful warm flush and the next snapshot did not go away. In this case, since we've already warm flushed, the loaded snapshot data would never get evicted.

Although... after revisiting this logic it makes more sense to me to just check during index bootstrapping whether or not the index block is sealed and we've successfully warm flushed and ignore any warm snapshots for that block.

Reply:
What do you think of this approach:
```
for blockStart, blockResults := range bootstrapResults {
        block, err := i.ensureBlockPresentWithRLock(blockStart.ToTime())
        if err != nil { // should never happen
                multiErr = multiErr.Add(i.unableToAllocBlockInvariantError(err))
                continue
        }
        // NB(bodu): For warm snapshots, we need to make sure that we haven't already successfully warm
        // flushed that block start. We can run into this case when the node crashes between a successful warm
        // flush and the next index snapshot.
        if _, ok := blockResults.GetBlock(idxpersist.SnapshotWarmIndexVolumeType); ok {
                if block.IsSealed() && i.hasIndexWarmFlushedToDisk(infoFiles, blockStart.ToTime()) {
                        // If we have warm snapshots and the block has been warm flushed already,
                        // we just discard the warm snapshot data.
                        blockResults.DeleteBlock(idxpersist.SnapshotWarmIndexVolumeType)
                }
        }
```

Reply:
+1